TestFoundation/TestNSNotificationQueue.swift

```diff
@@ -0,0 +1,151 @@
+        XCTAssertEqual(numberOfCalls, 1)
+    }
+
+    func test_postNowForDefaultRunLoopMode() {
+        let notificationName = "test_postNowToDefaultQueueWithCoalescingOnName"
+        let dummyObject = NSObject()
+        let notification = NSNotification(name: notificationName, object: dummyObject)
+        var numberOfCalls = 0
+        NSNotificationCenter.defaultCenter().addObserverForName(notificationName, object: dummyObject, queue: nil) { notification in
+            numberOfCalls += 1
+        }
+        let queue = NSNotificationQueue.defaultQueue()
+
+        let runLoop = NSRunLoop.currentRunLoop()
+        let endDate = NSDate(timeInterval: NSTimeInterval(0.05), sinceDate: NSDate())
+
+        let dummyTimer = NSTimer.scheduledTimer(0.01, repeats: false) { _ in
+            guard let runLoopMode = runLoop.currentMode else {
+                return
+            }
+
+            // post 2 notifications for the NSDefaultRunLoopMode mode
+            queue.enqueueNotification(notification, postingStyle: .PostNow, coalesceMask: [], forModes: [runLoopMode])
+            queue.enqueueNotification(notification, postingStyle: .PostNow)
+            // here we post notification for the NSRunLoopCommonModes. It shouldn't have any affect, because the timer is scheduled in NSDefaultRunLoopMode.
+            // The notification queue will only post the notification to its notification center if the run loop is in one of the modes provided in the array.
+            queue.enqueueNotification(notification, postingStyle: .PostNow, coalesceMask: [], forModes: [NSRunLoopCommonModes])
+        }
+        runLoop.addTimer(dummyTimer, forMode: NSDefaultRunLoopMode)
+        runLoop.runMode(NSDefaultRunLoopMode, beforeDate: endDate)
+        XCTAssertEqual(numberOfCalls, 2)
+    }
+
+    func test_notificationQueueLifecycle() {
+        // check that notificationqueue is associated with current thread. when the thread is destroyed, the queue should be deallocated as well
+        let defaultQueueForCurrentThread = NSNotificationQueue.defaultQueue()
+        weak var defaultQueueForBackgroundThread: NSNotificationQueue?
+        weak var notificationQueue: NSNotificationQueue?
+
+        self.executeInBackgroundThread() {
+            notificationQueue = NSNotificationQueue(notificationCenter: NSNotificationCenter())
+            XCTAssertNotNil(notificationQueue)
+            
+            defaultQueueForBackgroundThread = NSNotificationQueue.defaultQueue()
+            XCTAssertNotNil(defaultQueueForBackgroundThread)
+            XCTAssertNotEqual(defaultQueueForCurrentThread, defaultQueueForBackgroundThread!)
+        }
+
+        XCTAssertNil(defaultQueueForBackgroundThread)
+        XCTAssertNil(notificationQueue)
+    }
+
+    // MARK: Private
+
+    private func executeInBackgroundThread(operation: () -> ()) {
+        var isFinished = false
+        let bgThread = NSThread() {
+            operation()
+            NSThread.exit()
```

Comment:
direct calls to exit probably are not ideal here; the thread will exit on it's own accord when the block is finished.